### ISSUE TITLE
network instability issues (updated)

### ISSUE THREAD

**Issue by refacktor** [skepticoin/skepticoin#119]
These are fun times to be mining Skepticoin.

A few things I noticed:

- Only one peer remains up (not mine)
- My own nodes I try to run keep crashing, or they are running but don't appear on the list of peers when I try to add a 3rd or 4th machine.
- The one peer which is not mine behaves erratically, my logs showing Connection refused, Connection reset, No route to host.
- My node keeps retransmitting the entire blockchain to the one peer, over and over.
- Multiple "microforks" (blockchain diverges by few blocks) logged by the peers that manage to stay up a while.
- Additional instances of the one peer showing up on different TCP/IP Ports (non-2412) causing "No route to host" on my end
- Peers not responding to GetBlocksMessages or GetBlocksMessages not being sent. 
- Initially noticed GetBlocksMessages are sent twice to the same peer, after fixing this I noticed the peer disconnects in response to a legit 2nd GetBlocksMessages
- `netstat -an` showing "Recv-Q" backlog indicating the code is not reading the data from the sockets

**Comment by atramos** [comment]
Issues fixed in PR #120

**Comment by refacktor** [comment]
Updates:

- ~Only one peer remains up (not mine)~ and - ~My own nodes I try to run keep crashing, or they are running but don't appear on the list of peers when I try to add a 3rd or 4th machine.~ UPDATE: fixed, probably by #128 and/or #113
- ~The one peer which is not mine behaves erratically, my logs showing Connection refused, Connection reset, No route to host.~ UPDATE: Not sure if still happening, but hasn't been a problem
- My node keeps retransmitting the entire blockchain to the one peer, over and over. UPDATE: Possible fix in 6dd8a4c cherry-picked from withdrawn PR... currently I'm not running this code, not sure if it would cause other problems.
- ~Multiple "microforks" (blockchain diverges by few blocks) logged by the peers that manage to stay up a while.~ UPDATE: This is better now that my nodes are patched but won't be fully resolved until #129 is released and deployed to remaining nodes.
- Additional instances of the one peer showing up on different TCP/IP Ports (non-2412) causing "No route to host" on my end. UPDATE: Not sure about this, see possible fix in b1fcee2, cherry-picked from larger withdrawn PR, I'm not running this in prod.
- Peers not responding to GetBlocksMessages or GetBlocksMessages not being sent. UPDATE: I haven't checked on this recently but it was probably caused by the slow disk writes holding up the network thread, fixed in #129 
- Initially noticed GetBlocksMessages are sent twice to the same peer, after fixing this I noticed the peer disconnects in response to a legit 2nd GetBlocksMessages. UPDATE: Relates to c5f72b6 and 61ce589, from the withdrawn PR #120 - I ran this code for a while but then decided to roll it back for now.
- ~netstat -an showing "Recv-Q" backlog indicating the code is not reading the data from the sockets~ UPDATE: this was also caused by the slow disk writes -  #129 fixes this.
